### PR TITLE
ORC-191 Implement RLE v1 encoder

### PR DESCRIPTION
**Pull request by xndai**
Implement integer, byte and boolean RLE encoder, and add corresponding
UTs. Only v1 is implemented at this point.

Change-Id: Ia03b0d8bf98a843e45da8a5d349efeff9fb4fa00

### CODE REVIEW

**Comment by xndai** [comment]
@omalley @majetideepak Can you also take a look at this? Thanks.

**Comment by majetideepak** [comment]
@xndai I will take a look at this sometime tomorrow. Sorry for the delay.

File: c++/src/ByteRLE.cc [hunk 1]
```diff
@@ -27,6 +27,272 @@
...
+
+  ByteRleEncoderImpl::~ByteRleEncoderImpl() {
+    // PASS
+  }
+
+  void ByteRleEncoderImpl::writeByte(char c) {
+    if (bufferPosition == bufferLength) {
+      int addedSize = 0;
+      if (!outputStream->Next(reinterpret_cast<void **>(&buffer), &addedSize)) {
+        throw std::bad_alloc();
```
Comment (majetideepak) [inline c++/src/ByteRLE.cc]: When does `BufferedOutputStream::Next` return `false`?

File: c++/src/RLEv1.cc [hunk 2]
```diff
@@ -26,8 +26,173 @@
...
+const int MAX_DELTA = 127;
+const int MIN_DELTA = -128;
+const int MAX_LITERAL_SIZE = 128;
+
+RleEncoderV1::RleEncoderV1(
+                          std::unique_ptr<BufferedOutputStream> outStream,
+                          bool hasSigned):
+                          outputStream(std::move(outStream)) {
+  isSigned = hasSigned;
+  literals = new int64_t[MAX_LITERAL_SIZE];
```
Comment (majetideepak) [inline c++/src/RLEv1.cc]: use `std::vector<int64_t>` literals_buf as a buffer instead.

File: c++/src/RLEv1.hh [hunk 3]
```diff
@@ -26,6 +26,59 @@
...
+    std::unique_ptr<BufferedOutputStream> outputStream;
+    bool isSigned;
+    int64_t* literals;
+    int numLiterals;
+    int64_t delta;
+    bool repeat;
+    int tailRunLength;
+    int bufferPosition;
+    int bufferLength;
+    char * buffer;
```
Comment (majetideepak) [inline c++/src/RLEv1.hh]: `char*` (remove space)

File: c++/test/TestByteRLEEncoder.cc [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                       uint64_t numValues,
+                       char* notNull) {
+
+    std::unique_ptr<SeekableInputStream> inStream(
+      new SeekableArrayInputStream(memStream.getData(), memStream.getLength()));
+
+    std::unique_ptr<ByteRleDecoder> decoder =
+      createByteRleDecoder(std::move(inStream));
+
+    char* decodedData = new char[numValues];
```
Comment (majetideepak) [inline c++/test/TestByteRLEEncoder.cc]: use `std::vector<char> data_buf` as a buffer.

File: c++/test/TestByteRLEEncoder.cc [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        }
+      }
+    }
+  }
+
+  char * generateData(uint64_t numValues,
+                      uint64_t numNulls = 0,
+                      char ** notNull = nullptr) {
+    generateNotNull(numValues, numNulls, notNull);
+    char * res = new char[numValues];
```
Comment (majetideepak) [inline c++/test/TestByteRLEEncoder.cc]: pass `res` as an argument. Let's allocate and deallocate at the same place.

File: c++/test/TestByteRLEEncoder.cc [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+namespace orc {
+
+  const int DEFAULT_MEM_STREAM_SIZE = 1024 * 1024; // 1M
+
+  void generateNotNull(uint64_t numValues,
+                       uint64_t numNulls,
+                       char ** notNull) {
+    if (numNulls != 0 && notNull != nullptr) {
+      *notNull = new char[numValues];
```
Comment (majetideepak) [inline c++/test/TestByteRLEEncoder.cc]: pass `notNull` as an argument. Let's allocate and deallocate at the same place.

File: c++/test/TestByteRLEEncoder.cc [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        res[i] = static_cast<char>(std::rand() % 256);
+    }
+    return res;
+  }
+
+  char * generateBoolData(uint64_t numValues,
+                          uint64_t numNulls = 0,
+                          char ** notNull = nullptr) {
+    generateNotNull(numValues, numNulls, notNull);
+    char * res = new char[numValues];
```
Comment (majetideepak) [inline c++/test/TestByteRLEEncoder.cc]: pass `res` as an argument.

File: c++/test/TestByteRLEEncoder.cc [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    uint64_t block = 1024;
+    BufferedOutputStream bufStream(*pool, &memStream, capacity, block);
+
+    std::unique_ptr<BufferedOutputStream> outStream(
+        new BufferedOutputStream(*pool, &memStream, capacity, block));
+
+    std::unique_ptr<ByteRleEncoder> encoder =
+      createByteRleEncoder(std::move(outStream));
+
+    char* data = generateData(102400);
```
Comment (majetideepak) [inline c++/test/TestByteRLEEncoder.cc]: pass a `std::vector<char>` buffer.

File: c++/test/TestByteRLEEncoder.cc [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    BufferedOutputStream bufStream(*pool, &memStream, capacity, block);
+
+    std::unique_ptr<BufferedOutputStream> outStream(
+        new BufferedOutputStream(*pool, &memStream, capacity, block));
+
+    std::unique_ptr<ByteRleEncoder> encoder =
+      createByteRleEncoder(std::move(outStream));
+
+    char* notNull = nullptr;
+    char* data = generateData(102400, 377, &notNull);
```
Comment (majetideepak) [inline c++/test/TestByteRLEEncoder.cc]: allocate via `std::vector<char>`

File: c++/src/ByteRLE.cc [hunk 10]
```diff
@@ -27,6 +27,272 @@
...
+
+    void writeByte(char c);
+    void writeValues();
+    void write(char c);
+  };
+
+  ByteRleEncoderImpl::ByteRleEncoderImpl(
+                                std::unique_ptr<BufferedOutputStream> output)
+                                  : outputStream(std::move(output)) {
+    literals = new char[MAX_LITERAL_SIZE];
```
Comment (xndai) [inline c++/src/ByteRLE.cc]: I use unique_ptr here to indicate ByteRleEncoderImpl has full ownership of this buffer. Thanks for catching that.

File: c++/src/ByteRLE.cc (hunk 1 above)
Comment (xndai) [inline c++/src/ByteRLE.cc]: Next() is a method we inherit from protobuf ZeroCopyOutputStream. Current implementation of BufferedOutputStream doesn't return false, but we still want this check to comply with the function signature.

File: c++/src/ByteRLE.cc [hunk 11]
```diff
@@ -27,6 +27,272 @@
...
+    /**
+     * Flushing underlying BufferedOutputStream
+    */
+    virtual uint64_t flush() override;
+
+    virtual void recordPosition(PositionRecorder* recorder) const override;
+
+  protected:
+    std::unique_ptr<BufferedOutputStream> outputStream;
+    char * literals;
```
Comment (xndai) [inline c++/src/ByteRLE.cc]: done.

File: c++/src/ByteRLE.cc [hunk 12]
```diff
@@ -27,6 +27,272 @@
...
+    int numLiterals;
+    bool repeat;
+    int tailRunLength;
+    int bufferPosition;
+    int bufferLength;
+    char * buffer;
```
Comment (xndai) [inline c++/src/ByteRLE.cc]: done.

File: c++/src/ByteRLE.cc (hunk 10 above)
Comment (xndai) [inline c++/src/ByteRLE.cc]: thanks for catching that. I add a free() call. Regarding to RAII, I think it's a bit overkilled here. The array is guaranteed to be allocated in constructor and freed in destructor. The logic is quite clear.

File: c++/src/RLEv1.cc (hunk 2 above)
Comment (xndai) [inline c++/src/RLEv1.cc]: Same as above.

File: c++/src/RLEv1.hh (hunk 3 above)
Comment (xndai) [inline c++/src/RLEv1.hh]: done

File: c++/test/TestByteRLEEncoder.cc (hunk 5 above)
Comment (xndai) [inline c++/test/TestByteRLEEncoder.cc]: I have fixed these as well as the same issues in TestRLEv1Encoder.cc. See my next iteration.

File: c++/test/TestByteRLEEncoder.cc (hunk 4 above)
Comment (xndai) [inline c++/test/TestByteRLEEncoder.cc]: I don't see a huge difference here - we are manipulating raw buffer anyway. I would just leave them there unless you feel strongly about it.

Comment (majetideepak) [inline c++/test/TestByteRLEEncoder.cc]: Since this is only in the tests, let's leave it as is for now. Thanks.

**Comment by xndai** [comment]
Thanks @majetideepak  for reviewing. @prasanthj @omalley please help review and/or accept the PR. Thx!

**Comment by prasanthj** [comment]
Merged with master. Thanks @xndai  for the patch and @majetideepak  for the reviews!

**Pull request closed without merging**